text/0001-remove_ambiguous_grammar.md

```diff
@@ -0,0 +1,258 @@
+
+Consider the following query:
+
+```sql
+SELECT count FROM table GROUP count
+```
+
+Currently the values from this select will be grouped by the field key. But where we to every introduce some feature which would require the `COUNT` keyword after `GROUP` then this code will quietly change meaning.
+
+# 3. Proposal
+
+## Location specific reserved words.
+
+The first change proposed is to disallow raw identifiers in places where the same identifier could also be parsed as a keyword. For example, the following code would no longer be allowed:
+
+```sql
+use = 32;
+```
+
+`use` in this case could be the start of a USE statement in this position and would therefore not be allowed. In order to get the same code one would need to surround the identifier.
+
+```sql
+`use` = 32;
+```
+
+Another example would be:
+
+```sql
+SELECT field FROM use;
+```
+
+The part after `FROM` can also have statements, however the USE statement is disallowed in subqueries, therefore allowing `use` would not be a problem.
+
+This change would completely remove existing ambiguity wherever an expression could possibly be parsed as a statement. It would simplify the parser a great deal. This changes also does not require that raw identifiers are completely distinct.
+
+## Disallow raw identifiers to start with a digit
+
+This is a common way to distinguish between a digit and number. By disallowing the first character from being a digit we can be sure that Identifiers aren’t numbers. This would resolve the same text from being used as an identifier or a number depending on the place in the grammar.
+
+## Introduce strand prefixes
+
+The only way can currently be distinguished is if one strand fails to parse as any of the other types. This can lead to problems where a use wanted a certain value to be a plain strand but it happened to match another type of strand and is thus converted into that specific type.
+
+And example is `"5:00"` which brought up in an issue. The user wanted to store this value as a plain strand but it happened to match a thing strand.
+
+I propose we introduce specific strand prefixes for specific strand types:
+
+- `t"2012-05-26T12-00-00Z"` for date-time strands
+- `u'7c3f4ce8-83c4-458f-b1d7-b28352dea93c'` for UUID’s
+- `r"5:00"` for record id strings
+
+## Syntax Error for Block-RecordId Object ambiguity
+
+The following statement is ambiguous:
+
+```sql
+{ a:b.c }
+```
+
+This can be parsed either as an object with a field `a` and value `b.c`, or as a block statement taking the field `c` of the record `a:b`. Removing this ambiguity completely would require making significant changes to how record IDs are joined (changing the `:` to something else) or requiring that either blocks or objects always be enclosed by parentheses. Both of these changes are quite drastic for a relatively minor ambiguity.
```

Comment:
Is this a minor ambiguity? It seems like you can get completely different results from the same syntax? Or am I understanding wrong?

Reply:
I called it minor, because I don't think there are a lot of situation in which this ambiguity happens, but you are correct that you can get completely different results for the same syntax.